spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -158,6 +144,43 @@ case class RewriteMergeInto(spark: SparkSession) extends Rule[LogicalPlan] with
     }
   }
 
+  // when there are no matched actions, use a left anti join to remove any matching rows and rewrite to use
+  // append instead of replace. only unmatched source rows are passed to the merge and actions are all inserts.
+  private def buildInsertOnlyMergePlan(
+      targetTableScan: LogicalPlan,
+      targetTableOutput: Seq[AttributeReference],
+      source: LogicalPlan,
+      cond: Expression,
+      notMatchedActions: Seq[MergeAction]): LogicalPlan = {
+
+    notMatchedActions match {
+      case Seq(insertAction: InsertAction) =>
+        val filteredSource = insertAction.condition match {
```

Comment:
In case of a filter in `NOT MATCHED` clause, we can execute it before the join.

Reply:
Expressions in NOT MATCHED may only refer to the source relation columns.

Reply:
Do we have a test that validates this assumption?

Reply:
This is purely a Spark responsibility but I added a test too. Resolving this.